### DISCUSSION TITLE
Block ip address (in Traefik?)

### DISCUSSION THREAD

**Discussion by jankeesvw** [basecamp/kamal#926] [General]
I'm currently "under attack" by someone who is hammering my server, I tried using UFW but it seems that traffic isn't going through the firewall. Is there a way in Kamal to block a certain IP address when it misbehaves?

**Comment by scart88** [comment]
You can try https://github.com/rack/rack-attack or the new Rails `rate_limiter`
``` 
class PostsController < ApplicationController
  RATE_LIMIT_STORE = ActiveSupport::Cache::RedisCacheStore.new(url: ENV["REDIS_URL"])
  rate_limit to: 100, within: 1.minute, store: RATE_LIMIT_STORE, by: -> { client_ip }, only: :index
  
  ...
  
  def client_ip
    request.headers["HTTP_X_REQUEST_ID"] or whatever IP you can get
  end
```

Alternatively, you can use https://github.com/fail2ban/fail2ban

**Reply by jankeesvw** [reply]
Thanks, I'm already using `rate_limit`. But the CPU's where completely maxed out by Traefik, so even before it touched my Rails app.

**Comment by nickhammond** [comment]
@jankeesvw You might also look at options at your hosting provider(security group, virtual firewall), you could block all requests from the source IP.

**Reply by jankeesvw** [reply]
I'm using DigitalOcean, but they don't offer any options to block specific IPs. I ended up resorting to Cloudflare. See this comment: [https://github.com/basecamp/kamal/discussions/926#discussioncomment-10584219](https://github.com/basecamp/kamal/discussions/926#discussioncomment-10584219).

**Comment by dhh** [comment]
I can also highly recommend Cloudflare for DDoS protection, although if your origin IP is already under attack, maybe that's a little late.

**Reply by jankeesvw** [reply]
I installed Cloudflare as a nameserver, but the attacker was still connecting directly to the origin IP. I resolved this by creating a new droplet and setting it up – which only took a few minutes thanks to Kamal. 

It would be great if Kamal had a solution for this. I'd prefer not to have Cloudflare sitting between my server and my users. How do you handle this at 37Signals, @dhh? 

But I'm glad the attack is over now.

**Reply by dhh** [reply]
There's nothing that'll protect a single droplet from a DDoS attack, unfortunately. You need a monster pipe and sophisticated network gear. That's why Cloudflare is so popular.

Thankfully DDoS attacks are generally quite rare. We've only had a couple in 20+ years. But when they happen, you need a big, mean network brute to handle it. Cloudflare is that.
